File: .github/scripts/validate-publication-catalog.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "pathname"

module PublicationCatalog
  class ValidationError < StandardError; end

  Document = Struct.new(:id, :source, :optional, :home, keyword_init: true)

  ROOT_KEYS = %w[documents schema_version].freeze
  DOCUMENT_KEYS = %w[home id optional source].freeze
  ID_PATTERN = /\A[a-z0-9]+(?:-[a-z0-9]+)*\z/

  module_function

  def validate(catalog_path, root: Dir.pwd)
    root_path = Pathname.new(root).expand_path
    raise ValidationError, "Repository root does not exist: #{root_path}" unless root_path.directory?

    catalog = read_catalog(Pathname.new(catalog_path))
    validate_exact_keys(catalog, ROOT_KEYS, "publication catalog")

    unless catalog["schema_version"].is_a?(Integer) && catalog["schema_version"] == 1
      raise ValidationError, "publication catalog schema_version must be 1 and use an integer JSON value"
    end

    raw_documents = catalog["documents"]
    unless raw_documents.is_a?(Array) && !raw_documents.empty?
      raise ValidationError, "publication catalog documents must be a non-empty array"
    end

    documents = raw_documents.each_with_index.map do |raw_document, index|
      parse_document(raw_document, index, root_path)
    end

    duplicate_id = duplicate_value(documents.map(&:id))
    raise ValidationError, "Duplicate publication document id: #{duplicate_id}" if duplicate_id

    duplicate_source = duplicate_value(documents.map(&:source))
    if duplicate_source
      raise ValidationError, "Duplicate publication document source: #{duplicate_source}"
    end

    home_documents = documents.select(&:home)
    unless home_documents.length == 1
      raise ValidationError, "publication catalog must select exactly one home document"
    end
    if home_documents.first.optional
      raise ValidationError, "publication catalog home document must not be optional"
    end

    documents
  end

  def read_catalog(path)
    if path.symlink?
      raise ValidationError, "Publication catalog must not be a symlink: #{path}"
    end

    content = path.binread.force_encoding(Encoding::UTF_8)
    unless content.valid_encoding?
      raise ValidationError, "Invalid publication catalog JSON #{path}: content is not valid UTF-8"
    end

    JSON.parse(content)
  rescue Errno::ENOENT
    raise ValidationError, "Publication catalog does not exist: #{path}"
  rescue Errno::EACCES => e
    raise ValidationError, "Unable to read publication catalog #{path}: #{e.message}"
  rescue JSON::ParserError => e
    raise ValidationError, "Invalid publication catalog JSON #{path}: #{e.message}"
  end

  def parse_document(raw_document, index, root_path)
    field = "documents[#{index}]"
    unless raw_document.is_a?(Hash)
      raise ValidationError, "#{field} must be an object"
    end

    validate_exact_keys(raw_document, DOCUMENT_KEYS, field)

    id = raw_document["id"]
    unless id.is_a?(String) && ID_PATTERN.match?(id)
      raise ValidationError, "#{field}.id must use lowercase kebab-case"
    end

    source = validate_source(raw_document["source"], "#{field}.source")
    optional = raw_document["optional"]
    home = raw_document["home"]
    unless optional == true || optional == false
      raise ValidationError, "#{field}.optional must be boolean"
    end
    unless home == true || home == false
      raise ValidationError, "#{field}.home must be boolean"
    end

    validate_source_file(root_path, source, field)
    Document.new(id: id, source: source, optional: optional, home: home)
  end

  def validate_exact_keys(value, expected, field)
    unless value.is_a?(Hash)
      raise ValidationError, "#{field} must be an object"
    end

    actual = value.keys.sort
    return if actual == expected.sort

    missing = expected - value.keys
    unknown = value.keys - expected
    details = []
    details << "missing: #{missing.sort.join(', ')}" unless missing.empty?
    details << "unsupported: #{unknown.sort.join(', ')}" unless unknown.empty?
    raise ValidationError, "#{field} fields are invalid (#{details.join('; ')})"
  end

  def validate_source(value, field)
    unless value.is_a?(String) && !value.empty?
      raise ValidationError, "#{field} must be a non-empty string"
    end

    parts = value.split("/", -1)
    unsafe = value.start_with?("/") || value.include?("\\") || value.include?("\0") ||
             parts.any? { |part| part.empty? || part == "." || part == ".." }
    raise ValidationError, "#{field} must be a safe relative POSIX path: #{value.inspect}" if unsafe

    unless value.downcase.end_with?(".md")
      raise ValidationError, "#{field} must identify a Markdown file"
    end

    value
  end

  def validate_source_file(root_path, source, field)
    candidate = root_path
    source.split("/").each do |part|
      candidate = candidate.join(part)
      if candidate.symlink?
        raise ValidationError, "#{field} traverses a symlink: #{source}"
      end
    end

    unless candidate.file?
      raise ValidationError, "#{field} does not identify an existing regular file: #{source}"
    end
  end

  def duplicate_value(values)
    seen = {}
    values.each do |value|
      return value if seen[value]

      seen[value] = true
    end
    nil
  end
end

if $PROGRAM_NAME == __FILE__
  root = Pathname.new(ARGV[1] || Dir.pwd).expand_path
  catalog_path = ARGV[0] ? Pathname.new(ARGV[0]) : root.join("docs/publication-catalog.json")

  begin
    documents = PublicationCatalog.validate(catalog_path, root: root)
  rescue PublicationCatalog::ValidationError => e
    warn "validate-publication-catalog.rb: #{e.message}"
    exit 1
  end

  puts "Publication catalog valid: #{documents.length} document(s)."
end

File: .github/scripts/validate_publication_catalog.py
#!/usr/bin/env python3
"""Validate the Skill documentation publication catalog."""

from __future__ import annotations

import json
import os
import re
import stat
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any


ROOT_KEYS = {"documents", "schema_version"}
DOCUMENT_KEYS = {"home", "id", "optional", "source"}
ID_PATTERN = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


class ValidationError(Exception):
    """Raised when the publication catalog violates its contract."""


@dataclass(frozen=True)
class Document:
    id: str
    source: str
    optional: bool
    home: bool


def _validate_exact_keys(value: Any, expected: set[str], field: str) -> None:
    if not isinstance(value, dict):
        raise ValidationError(f"{field} must be an object")

    actual = set(value)
    if actual == expected:
        return

    details: list[str] = []
    missing = sorted(expected - actual)
    unknown = sorted(actual - expected)
    if missing:
        details.append(f"missing: {', '.join(missing)}")
    if unknown:
        details.append(f"unsupported: {', '.join(unknown)}")
    raise ValidationError(f"{field} fields are invalid ({'; '.join(details)})")


def _read_catalog(path: Path) -> Any:
    if path.is_symlink():
        raise ValidationError(f"Publication catalog must not be a symlink: {path}")

    try:
        raw = path.read_bytes()
    except FileNotFoundError as exc:
        raise ValidationError(f"Publication catalog does not exist: {path}") from exc
    except PermissionError as exc:
        raise ValidationError(f"Unable to read publication catalog {path}: {exc}") from exc

    try:
        content = raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise ValidationError(
            f"Invalid publication catalog JSON {path}: content is not valid UTF-8"
        ) from exc

    try:
        return json.loads(content)
    except json.JSONDecodeError as exc:
        raise ValidationError(f"Invalid publication catalog JSON {path}: {exc}") from exc


def _validate_source(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value:
        raise ValidationError(f"{field} must be a non-empty string")

    parts = value.split("/")
    unsafe = (
        value.startswith("/")
        or "\\" in value
        or "\x00" in value
        or any(part in {"", ".", ".."} for part in parts)
    )
    if unsafe:
        raise ValidationError(f"{field} must be a safe relative POSIX path: {value!r}")

    if not value.lower().endswith(".md"):
        raise ValidationError(f"{field} must identify a Markdown file")

    return value


def _validate_source_file(root: Path, source: str, field: str) -> None:
    candidate = root
    for part in source.split("/"):
        candidate = candidate / part
        if candidate.is_symlink():
            raise ValidationError(f"{field} traverses a symlink: {source}")

    try:
        mode = candidate.stat().st_mode
    except FileNotFoundError as exc:
        raise ValidationError(
            f"{field} does not identify an existing regular file: {source}"
        ) from exc

    if not stat.S_ISREG(mode):
        raise ValidationError(
            f"{field} does not identify an existing regular file: {source}"
        )


def _parse_document(raw_document: Any, index: int, root: Path) -> Document:
    field = f"documents[{index}]"
    _validate_exact_keys(raw_document, DOCUMENT_KEYS, field)

    document_id = raw_document["id"]
    if not isinstance(document_id, str) or ID_PATTERN.fullmatch(document_id) is None:
        raise ValidationError(f"{field}.id must use lowercase kebab-case")

    source = _validate_source(raw_document["source"], f"{field}.source")
    optional = raw_document["optional"]
    home = raw_document["home"]
    if type(optional) is not bool:
        raise ValidationError(f"{field}.optional must be boolean")
    if type(home) is not bool:
        raise ValidationError(f"{field}.home must be boolean")

    _validate_source_file(root, source, field)
    return Document(id=document_id, source=source, optional=optional, home=home)


def _duplicate_value(values: list[str]) -> str | None:
    seen: set[str] = set()
    for value in values:
        if value in seen:
            return value
        seen.add(value)
    return None


def validate(
    catalog_path: str | os.PathLike[str],
    root: str | os.PathLike[str] = ".",
) -> list[Document]:
    root_path = Path(root).expanduser().resolve()
    if not root_path.is_dir():
        raise ValidationError(f"Repository root does not exist: {root_path}")

    resolved_catalog_path = Path(catalog_path).expanduser()
    if not resolved_catalog_path.is_absolute():
        resolved_catalog_path = root_path / resolved_catalog_path

    catalog = _read_catalog(resolved_catalog_path)
    _validate_exact_keys(catalog, ROOT_KEYS, "publication catalog")

    schema_version = catalog["schema_version"]
    if type(schema_version) is not int or schema_version != 1:
        raise ValidationError(
            "publication catalog schema_version must be 1 and use an integer JSON value"
        )

    raw_documents = catalog["documents"]
    if not isinstance(raw_documents, list) or not raw_documents:
        raise ValidationError("publication catalog documents must be a non-empty array")

    documents = [
        _parse_document(raw_document, index, root_path)
        for index, raw_document in enumerate(raw_documents)
    ]

    duplicate_id = _duplicate_value([document.id for document in documents])
    if duplicate_id is not None:
        raise ValidationError(f"Duplicate publication document id: {duplicate_id}")

    duplicate_source = _duplicate_value([document.source for document in documents])
    if duplicate_source is not None:
        raise ValidationError(f"Duplicate publication document source: {duplicate_source}")

    home_documents = [document for document in documents if document.home]
    if len(home_documents) != 1:
        raise ValidationError("publication catalog must select exactly one home document")
    if home_documents[0].optional:
        raise ValidationError("publication catalog home document must not be optional")

    return documents


def main(argv: list[str] | None = None) -> int:
    args = list(sys.argv[1:] if argv is None else argv)
    if len(args) > 2:
        print(
            f"usage: {Path(sys.argv[0]).name} [CATALOG_PATH] [REPOSITORY_ROOT]",
            file=sys.stderr,
        )
        return 2

    root = Path(args[1] if len(args) == 2 else Path.cwd()).expanduser().resolve()
    catalog_path = Path(args[0]) if args else root / "docs/publication-catalog.json"

    try:
        documents = validate(catalog_path, root=root)
    except ValidationError as exc:
        print(f"validate_publication_catalog.py: {exc}", file=sys.stderr)
        return 1

    print(f"Publication catalog valid: {len(documents)} document(s).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
